Fix long decimal partial aggregation below join

findepi opened the pull request:
Fixes https://github.com/trinodb/trino/issues/21099

findepi commented:
according to @lukasz-stec 's analysis, this may be fixing some bug related to partial aggregations, but i don't know how to reproduce that bug.

lukasz-stec reviewed (approved):
lgtm % comments

findepi commented on core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java:
```diff
@@ -74,30 +73,28 @@ public void deserialize(Block block, int index, LongDecimalWithOverflowAndLongSt
-        Slice slice = variableWidthBlock.getRawSlice();
-        int sliceOffset = variableWidthBlock.getRawSliceOffset(index);
-        int sliceLength = variableWidthBlock.getSliceLength(index);
+        Slice slice = VARBINARY.getSlice(block, index);
```
yeah, i don't know what the contract is for state serializers, that's why i didn't add any test

do we have an example I should follow?

dain commented on core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java:
```diff
@@ -74,30 +73,28 @@ public void deserialize(Block block, int index, LongDecimalWithOverflowAndLongSt
-        Slice slice = variableWidthBlock.getRawSlice();
-        int sliceOffset = variableWidthBlock.getRawSliceOffset(index);
-        int sliceLength = variableWidthBlock.getSliceLength(index);
+        Slice slice = VARBINARY.getSlice(block, index);
```
I like this solution. I also would not wory about the object allocation.

raunaqmorarka replied:
There was pretty clear performance impact of object allocation here https://github.com/trinodb/trino/pull/18868#issuecomment-1699603094
We should at least benchmark it before deciding either way.

dain replied:
Honestly, if allocating a single Slice (which is basically a record) is 10% of CPU, I'd expect that benchmark is broken  given how much other stuff is happening in an aggregation query.

dain replied:
Anyway, if you actually care, my initial suggestion was:
```
index = block.getUnderlyingValuePosition(index);
block = block.getUnderlyingValueBlock();

findepi replied:
> `block.getUnderlyingValueBlock` and `block.getUnderlyingValuePosition` can still be virtual calls so there is potential regression here anyway.

good point.

sopel39 reviewed (changes requested):
There is also:
* LongDecimalWithOverflowStateSerializer
* @lukasz-stec comments

Alternative approach is to handle dictionaries in `StateCompiler#generateDeserialize` as it would prevent polymorphic calls and extra indirection

findepi commented:
@lukasz-stec can you help me with testing this?
would especially be good to have query-based regression that that uncovers the necessity for state serializers to handle dict blocks.

lukasz-stec commented:
> @lukasz-stec can you help me with testing this? would especially be good to have query-based regression that that uncovers the necessity for state serializers to handle dict blocks.

@findepi I don't know how to repro this using a query. It may be imposiible at the current state. It requires partial and final or intermediate aggregation to be separated by operation that can produce dictionary blocks for aggregation state. Join is one potential such operation.

Here are unit tests that cover this though.
[Test_dict_and_rle_blocks_deserialization.patch](https://github.com/trinodb/trino/files/14613052/Test_dict_and_rle_blocks_deserialization.patch)

findepi commented:
[](url)

> There is also:
> 
> * LongDecimalWithOverflowStateSerializer

@sopel39 this is the first class i fixed. is there some other state serializer class i missed?
i searched for `(VariableWidthBlock)` in remaining state serializers, but didn't find any

findepi commented:
> I don't know how to repro this using a query. It may be imposiible at the current state

i was able -- https://github.com/trinodb/trino/issues/21099

findepi commented:
> Here are unit tests that cover this though.
> [Test_dict_and_rle_blocks_deserialization.patch](https://github.com/trinodb/trino/files/14613052/Test_dict_and_rle_blocks_deserialization.patch)

thanks!

findepi commented:
i added a unit test from @lukasz-stec , thank you
i added a query-based regression test (from https://github.com/trinodb/trino/issues/21099#issuecomment-1999197043)

There is open discussion how to write the code the most performant way and there seems to be disagreement how this should look like. I propose we merge the code, since it's an obvious improvement over current (broken) state.
Alternatively we do not merge this code, but then I would like you guys, @sopel39 @raunaqmorarka @lukasz-stec , take this over.

PTAL

sopel39 commented:
@findepi let's use @dain approach:
https://github.com/trinodb/trino/pull/21083/commits/112788b299536e94c6554803e1de19149aa62418#r1525466731
if you don't plan to handle dictionaries in `StateCompiler#generateDeserialize`

findepi commented:
@lukasz-stec had concerns. @lukasz-stec  will you be OK with that approach?

lukasz-stec commented:
> @lukasz-stec had concerns. @lukasz-stec will you be OK with that approach?

Yes, the only concern was a potential virtual call and a) it is not that bad, b) it will happen only in rare conditions if at all as it needs all three block types to be there (value, dict, RLE)

findepi commented:
updated, PTAL

sopel39 reviewed (approved):
awesome! thanks

mosabua commented:
Since this fixes an issue I added a release notes entry
